django/db/models/deletion.py

```diff
@@ -247,7 +281,31 @@ def collect(self, objs, source=None, nullable=False, collect_related=True,
                 if hasattr(field, 'bulk_related_objects'):
                     # It's something like generic foreign key.
                     sub_objs = field.bulk_related_objects(new_objs, self.using)
-                    self.collect(sub_objs, source=model, nullable=True)
+                    self.collect(sub_objs, source=model, nullable=True, fail_on_restricted=False)
+
+        if fail_on_restricted:
+            # During collection we may have found objects via RESTRICT
+            # relationships (restricted objects). To be able to delete them, we
+            # need to also have found them through a different CASCADE
+            # relationship. Fail here if any of them wasn't.
+            for model, instances in self.data.items():
+                self.clear_restricted_objects_from_set(model, instances)
```

Comment:
Do we need this? I can comment these two lines and all tests pass.

Reply:
Added test `test_restrict_generic_foreign_key_and_no_fast_delete()` to exercise these lines. They're needed whenever a model protected by `RESTRICT` can't be "fast deleted" as per `collect() internals. Confirmed that removing the lines makes it fail.

Also made this new test use generic foreign keys because no existing test uses them.